File: quick_sort_step.java

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class quick_sort_step {

    static class Data {

        int number;
        String text;

        Data(int number, String text) {
            this.number = number;
            this.text = text;
        }

        @Override
        public String toString() {
            return number + "/" + text; // format: "number/text"
        }
    }

    public static void main(String[] args) throws IOException {
        String inputFile = "dataset_sample_1000.csv";
        int startRow = 1;
        int endRow = 7;

        List<Data> list = new ArrayList<>();

        // Reads text from the csv file
        try (BufferedReader br = new BufferedReader(new FileReader(inputFile))) {
            int currentLine = 1; //1-based
            String line;

            //read the file line by line
            while ((line = br.readLine()) != null) {
                if (currentLine >= startRow && currentLine <= endRow) {
                    String[] parts = line.split(","); //split csv file
                    list.add(new Data(Integer.parseInt(parts[0]), parts[1]));
                }
                currentLine++; //move next linee
            }
        }

        // Writes the sorting steps to output file
        try (BufferedWriter bw = new BufferedWriter(
                new FileWriter("quick_sort_step_" + startRow + "_" + endRow + ".txt"))) { //filename
            printList(bw, list); // write original array (unsorted array)
            quickSort(list, 0, list.size() - 1, bw); // start sort and log
        }
    }

    //-----Quicksort implementation--------------------- (inside lecture notes)
    static void quickSort(List<Data> arr, int left, int right, BufferedWriter bw) throws IOException {
        if (left < right) { //only sort if atleast got 2 elements

            int pi = partition(arr, left, right); //get pivot index position
            printList(bw, arr, pi); // log current state with pivot

            quickSort(arr, left, pi - 1, bw);    //left side of pivot
            quickSort(arr, pi + 1, right, bw);   //right side of pivot
        }
    }

    //-----Partition for quicksort--------------------------(lecture notes use LinkedList)
    static int partition(List<Data> arr, int left, int right) {
        // Create temporary queues
        LinkedList<Data> queue = new LinkedList<>(arr.subList(left, right + 1));

        LinkedList<Data> L = new LinkedList<>(); // Elements less than pivot
        LinkedList<Data> E = new LinkedList<>(); // Elements equal to pivot
        LinkedList<Data> G = new LinkedList<>(); // Elements greater than pivot

        Data pivot = queue.getLast(); // Pivot is last element

        while (!queue.isEmpty()) {
            Data current = queue.removeFirst();
            if (current.number < pivot.number) {
                L.addLast(current);
            } else if (current.number == pivot.number) {
                E.addLast(current);
            } else {
                G.addLast(current);
            }
        }

        // Rebuild the original list
        int index = left;
        for (Data d : L) {
            arr.set(index++, d); // Copy all 'less than' elements back

                }int pivotIndex = index;
        for (Data d : E) {
            arr.set(index++, d); //remeber where pivot section starts

                }for (Data d : G) {
            arr.set(index++, d); // Copy all than elements back
        }
        return pivotIndex;
    }

    static void printList(BufferedWriter bw, List<Data> list) throws IOException {
        for (Data d : list) {
            bw.write(d.toString() + ", ");
        }
        bw.newLine();
    }

    static void printList(BufferedWriter bw, List<Data> list, int pivotIndex) throws IOException {
        bw.write("pi=" + pivotIndex + " [");
        for (int i = 0; i < list.size(); i++) {
            bw.write(list.get(i).toString());
            if (i < list.size() - 1) {
                bw.write(", ");
            }
        }
        bw.write("]");
        bw.newLine();
    }
}

File: quick_sort_step.py
import csv
from collections import deque # For using queue-like linked lists

#----Class to represent each record (number, text)----
class Data:
    def __init__(self, number, text):
        self.number = number
        self.text = text

    def __str__(self):
        return f"{self.number}/{self.text}"

#----Read only a portion of the CSV file based on start and end row----
def read_csv_subset(filename, start_row, end_row):
    data_list = []
    with open(filename, 'r') as csvfile:
        reader = csv.reader(csvfile)
        for i, row in enumerate(reader, start=1):     # 1-based index
            if start_row <= i <= end_row:
                number = int(row[0])                  # Parse number
                text = row[1]                         # Get associated string
                data_list.append(Data(number, text))  # Add as Data object
    return data_list        # Return list of selected Data


#----Write the current list state to output file----
def write_list(bw, data_list, pivot_index=None):
    if pivot_index is None:         # the First line: print unsorted list
        bw.write(", ".join(str(d) for d in data_list)) # Convert each item to string
        bw.write("\n")
    else:     # Print list during sorting with pivot index
        line = f"pi={pivot_index} ["
        line += ", ".join(str(d) for d in data_list)
        line += "]\n"
        bw.write(line) # Write full sorting step

#----Partition method using deque----
def partition(arr, left, right):
    # Create a temporary queue containing the sublist to sort
    queue = deque(arr[left:right+1])

    L = deque()  # Elements less than pivot
    E = deque()  # Elements equal to pivot
    G = deque()  # Elements greater than pivot

    pivot = queue[-1] # Pivot is the last element

#-- Partition method : distribute elements into L, E, G
    while queue:                                   # While queue is not empty
        current = queue.popleft()                  # Get first element
        if current.number < pivot.number:
            L.append(current)                      # Add to less than pivot
        elif current.number == pivot.number:
            E.append(current)                      # Add to equal
        else:
            G.append(current)                      # Add to greater than pivot

    #-- Copy the partitions back into the original array
    index = left          # Start from left position
    for d in L:
        arr[index] = d    # Set element in original list
        index += 1
    pivot_index = index   # Mark where pivot group begins
    for d in E:
        arr[index] = d
        index += 1
    for d in G:
        arr[index] = d
        index += 1

    return pivot_index   # Return the new pivot index for further sorting

#----Quick Sort recursive implementation ----
def quick_sort(arr, left, right, bw):
    if left < right:
        pi = partition(arr, left, right)  # Partition and get pivot index (above code)
        write_list(bw, arr, pi)           # Write current state with pivot index
        quick_sort(arr, left, pi - 1, bw) #sort left side
        quick_sort(arr, pi + 1, right, bw) #sort right side

#----Main driver function----
def main():
    input_file = "dataset_sample_1000.csv"
    start_row = 1
    end_row = 7
    output_file = f"quick_sort_step_{start_row}_{end_row}.txt"

    data_list = read_csv_subset(input_file, start_row, end_row)

    with open(output_file, "w") as bw:
        write_list(bw, data_list)  # Initial unsorted list
        quick_sort(data_list, 0, len(data_list) - 1, bw)

if __name__ == "__main__":
    main()